File: algo-dsa-beginner/02_static-array/python/remove-ele.py
class Solution:

    def removeElement(self, nums: list[int], val:int) -> int:
        # lets inchworm l2r

        r_head = 0
        w_head = 0

        while (r_head < len(nums)):
            if (nums[r_head] != val):
                nums[w_head] = nums[r_head]
                w_head += 1

            r_head += 1

        del nums[w_head:]
        return w_head

    def removeElement2(self, nums: list[int], val: int) -> int:
        # could also do solution where stack allocate temp array to copy valid elements in,
        # and then copy back valid entries back into orignal array
        # or could force a reorder

        # depends on what we are optimizing for

        if len(nums) == 0:
            return 0

        lr_head = 0
        rl_head = len(nums) - 1

        #lr advances, if encounters val then swaps with rl_head
        #rl_head points to furthest avail valid val for swap
        #at the end rl_head will be meet lr head at end of valid array
        while lr_head < rl_head:
            if nums[rl_head] == val:
                rl_head -= 1
            elif nums[lr_head] == val:
                nums[lr_head] = nums[rl_head]
                rl_head -= 1
            else:
                lr_head += 1


        # forgot edge case if entire array is made up of value to remove
        return 0 if lr_head == 0 else lr_head + 1



ex1 = [1,1,2,3,4]
ex1_out = [2,3,4]
ex1_ret = 3

ex2 = [0,1,2,2,3,0,4,2]
ex2_out = [0,1,3,0,4]
ex2_ret = 5

sol = Solution()
print()

print(f"ex1: {ex1}, removing: 1")
ret = sol.removeElement(ex1, 1)
print(f"expected: {ex1_out}")
print(f"actual: {ex1}")
print(f"expected: {ex1_ret}")
print(f"ex1 ret = {ret}")
print()

print(f"ex2: {ex2}, removing: 2")
ret = sol.removeElement(ex2, 2)
print(f"expected: {ex2_out}")
print(f"actual: {ex2}")
print(f"expected: {ex2_ret}")
print(f"ex2 ret = {ret}")
print()

File: algo-dsa-beginner/02_static-array/cpp/remove-ele.cpp
#include <iostream>
#include <string>
#include <vector>

template <typename T> using vector = std::vector<T>;
using string = std::string;

template <typename T>
void print_vec(const vector<T>& nums) {
    std::cout << "[ ";
    for (size_t i = 0; i < nums.size(); i++) {
        std::cout << (i == 0 ? "" : ", ") << nums[i];
    }
    std::cout << " ]\n";
}

class Solution {

public:

int remove_element(vector<int>& nums, int val) {
    print_vec(nums);
    return val;
}

};


int main() {

    Solution sol;

    std::cout << "\nex1: ";
    vector<int> ex1 = {1,1,2,3,4};
    print_vec(ex1);

    auto ret = sol.remove_element(ex1, 1);
    std::cout << "ex1 expected: [2,3,4]\n";
    std::cout << "ex1 ret expected: 3, acutal: " << ret << "\n";

    std::cout << "\nex2: ";
    vector<int> ex2 = {0,1,2,2,3,0,4,2};
    print_vec(ex2);
    ret = sol.remove_element(ex2, 2);
    std::cout << "ex2 expected: [0,1,3,0,4]\n";
    std::cout << "ex2 ret expected: 5, acutal: " << ret << "\n";

    return 0;
}
